tests/test_compile_kwargs.py

```diff
@@ -0,0 +1,555 @@
+from inspect import isclass
+
+import numpy as np
+import pytest
+
+from numpy.lib.arraysetops import isin
+from sklearn.datasets import make_classification
+from tensorflow.keras import losses as losses_module
+from tensorflow.keras import metrics as metrics_module
+from tensorflow.keras import optimizers as optimizers_module
+from tensorflow.keras.layers import Dense, Input
+from tensorflow.keras.models import Model
+from tensorflow.python.keras.engine.compile_utils import (
+    losses_mod,
+    metrics_mod,
+)
+
+from scikeras.wrappers import KerasClassifier
+
+
+def get_model(num_hidden=10, meta=None, compile_kwargs=None):
+    inp = Input(shape=(meta["n_features_in_"],))
+    hidden = Dense(num_hidden, activation="relu")(inp)
+    out = [
+        Dense(1, activation="sigmoid", name=f"out{i+1}")(hidden)
+        for i in range(meta["n_outputs_"])
+    ]
+    model = Model(inp, out)
+    model.compile(**compile_kwargs)
+    return model
+
+
+@pytest.mark.parametrize("optimizer", (optimizers_module.SGD, "SGD", "sgd"))
+@pytest.mark.parametrize("n_outputs_", (1, 2))
+def test_optimizer(optimizer, n_outputs_):
+    """Tests compiling of single optimizer with options.
+    Since there can only ever be a single optimizer, there is no
+    ("name", optimizer, "output") option.
+    Only optimizer classes will be compiled with custom options,
+    all others (class names, function names) should pass through
+    untouched.
+    """
+    # Single output
+    X, y = make_classification()
+    y = np.column_stack([y for _ in range(n_outputs_)]).squeeze()
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer=optimizer,
+        optimizer__learning_rate=0.15,
+        optimizer__momentum=0.5,
+        loss="binary_crossentropy",
+    )
+    est.fit(X, y)
+    if isclass(optimizer):
+        np.testing.assert_almost_equal(
+            est.model_.optimizer.momentum.value().numpy(), 0.5, decimal=2
+        )
+        np.testing.assert_almost_equal(
+            est.model_.optimizer.learning_rate.value().numpy(), 0.15, decimal=2
+        )
+    else:
+        # TF actually instantiates an optimizer object even if passed a string
+        # Checking __class__ is not a universally valid check, but it works for the
+        # test optimizer (SGD)
+        assert (
+            est.model_.optimizer.__class__
+            == optimizers_module.get(optimizer).__class__
+        )
+
+
+def test_optimizer_invalid_string():
+    """Tests that a ValueError is raised when an unknown
+    string is passed as an optimizer.
+    """
+
+    X, y = make_classification()
+
+    optimizer = "sgf"  # sgf is not a loss
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer=optimizer,
+        loss="binary_crossentropy",
+    )
+    with pytest.raises(ValueError, match="Unknown optimizer"):
+        est.fit(X, y)
+
+
+def test_compiling_of_routed_parameters():
```

Comment:
This code raises "TypeError: model() got an unexpected keyword argument 'foo__bar'"

``` python
class Foo:
    def __init__(self, bar):
        self.bar = bar

class Bar:
    def __init__(self, x=3):
        self.x = x

def model(foo: Foo):
    inputs = tf.keras.Input(shape=(100,))
    x = tf.keras.layers.Dense(4, activation=tf.nn.relu)(inputs)
    outputs = tf.keras.layers.Dense(1, activation=tf.nn.softmax)(x)
    model = tf.keras.Model(inputs=inputs, outputs=outputs)
    return model

est = KerasRegressor(
    model=model,
    model__foo=Foo,
    model__foo__bar=Bar,
    model__foo__bar__x=5,
)
```

I don't think I'd expect that.

Reply:
Interesting but I think that's largely unrelated to this PR right? This is just parameter routing. You routed `foo_bar` to `model` but your `model` function does not accept a parameter by that name. SciKeras does not compile arbitrary user parameters, only `optimizer/loss/metrics`.

Reply:
Closing in favor of #82.